### ISSUE TITLE
Astro pre-evaluates dynamic imports that prevents tree-shaking

### ISSUE THREAD

**Issue by putnik-projects** [withastro/astro#6328]
### What version of `astro` are you using?

2.0.14

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

npm

### What operating system are you using?

It doesn't matter for this bug

### Describe the Bug

All the dynamic imports and `import.meta.glob` imports are bundled in my production build, regardless of whether they are used on the page. I could resolve all the modules by myself when needed, but it's impossible due to this bug. For a more detailed description of the bug and why I need this feature, see https://github.com/vitejs/vite/discussions/12104.

Here is another reproduction of the bug: https://stackblitz.com/edit/github-b1xdf7?file=src%2Fpages%2Findex.astro. I also think that this feature would help a lot of projects to significantly reduce the size of a bundle.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-b1xdf7-rajkhn?file=src%2Fpages%2Findex.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by rishi-raj-jain** [comment]
In your reproduction, if you comment out import '../components/dynamic/FirstComponentWithJS.astro' in about.astro, doesn't that work as expected?

**Comment by putnik-projects** [comment]
Yes, I can't just comment out unnecessary imports, because I don't know which ones are used on a page, since I use cms.

**Comment by wulinsheng123** [comment]
mark

**Comment by matthewp** [comment]
This looks like the same as https://github.com/withastro/astro/issues/3816 . Importing a component does result in its styles being added to the page. We decide what styles are added through static analysis, not at runtime.

Content collections solves this problem and if you use that, only the styles used are added. We won't be changing the behavior of Astro.glob or import.meta.glob (and can't, as that other issue describes).

**Comment by putnik-projects** [comment]
Ok, as far as I'm concerned, in Astro `import.meta.glob` always runs and builds scripts and styles in all imported components anyway. That's why you've created the Content Collections API. So if I want to execute astro components dynamically, I have to first import them into an mdx file and then import that mdx file into the page and call the `.render()` method. This doesn't look very concise and obvious, but that's the only option I have, since `import.meta.glob` and `import()` in astro work in a different way rather than in vite itself. Therefore, @matthewp, it might be worth pointing this out in the documentation, because [the current note](https://docs.astro.build/en/reference/api-reference/#astroglob) about vite's `glob` is misleading:

> You can also use import.meta.glob() itself in your Astro project. You may want to do this when:
> - ...
> - You don’t want to load each file immediately. import.meta.glob() can return functions that import the file content, rather than returning the content itself.
